### PR TITLE
backport: bitcoin#12134, #18426, #18534, #18828, #18864, #19153, #19201, #19205, #19560, #19813, #19859, partial #20354, #20955, #22442, #22790

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
That's the last batch of backports from bitcoin v20. Beside that there's some related backports from other versions and related fixes.

Note for reviewer: please, notice, that some particular commits are not dashified because later on theirs code is changed or removed in next backports of this PR. But final version of all changes is dashified (except wallet_upgradewallet.py that was backported some time ago in past and it still doesn't work).

## What was done?
Backports:
- Merge bitcoin/bitcoin#12134
- Merge bitcoin/bitcoin#18426
- Merge bitcoin/bitcoin#18534
- Merge bitcoin/bitcoin#18828
- Merge bitcoin/bitcoin#19201
- Merge bitcoin/bitcoin#19153
- Merge bitcoin/bitcoin#18864
- Merge bitcoin/bitcoin#19205
- Merge bitcoin/bitcoin#19560
- Merge bitcoin/bitcoin#19813
- Merge bitcoin/bitcoin#19859
- Merge bitcoin/bitcoin#20955
- Merge bitcoin/bitcoin#22442
- partial Merge bitcoin/bitcoin#20354
- Merge bitcoin/bitcoin#22790

Missing changes from other backports that were done before:
 - bitcoin/bitcoin#18067
 - bitcoin/bitcoin#18253
 - bitcoin/bitcoin#18774
 - bitcoin/bitcoin#18873
 - bitcoin/bitcoin#20034

All these backports are bringing next changes:
 - new functional test `mempool_compatibility.py`
 - new functional test `feature_backwards_compatibility.py` 
 - enables backported before functional test `wallet_upgradewallet.py`
 - utility script `test/get_previous_releases.py` to get binaries with previous version

Although, `wallet_upgradewallet.py` is disabled intentionally - it's not adopted for dash specific so far as HD wallets are not enabled by default atm.

## How Has This Been Tested?
Run unit/functional tests.
Beside that tested new functional script. Steps to do it:
```
$ test/get_previous_releases.py -b v19.3.0 v18.2.2 v0.17.0.3 v0.16.1.1 v0.15.0.0
$ test/functional/test_runner.py feature_backwards_compatibility.py mempool_compatibility.py wallet_upgradewallet.py
feature_backwards_compatibility.py | ✓ Passed  | 15 s
mempool_compatibility.py           | ✓ Passed  | 4 s
wallet_upgradewallet.py            | ○ Skipped | 1 s
ALL                                | ✓ Passed  | 20 s (accumulated) 
```

## Breaking Changes
N/A

## Checklist:
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by knst** [comment]
> [samettek](https://github.com/samettek) approved these changes

little too early, it was not ready for review yet

File: test/functional/wallet_upgradewallet.py [hunk 1]
```diff
@@ -26,15 +25,18 @@ def set_test_params(self):
...
             ["-usehd=0"]             # v0.15.2 wallet
         ]
+        self.wallet_names = [self.default_wallet_name]
 
     def skip_test_if_missing_module(self):
         self.skip_if_no_wallet()
         self.skip_if_no_bdb()
         self.skip_if_no_previous_releases()
+        # TODO: this test doesn't work yet
+        raise SkipTest("Test wallet_upgradewallet.py is not adopted for Dash Core yet.")
```
Comment (UdjinM6) [inline test/functional/wallet_upgradewallet.py:39]:
```suggestion
        raise SkipTest("Test wallet_upgradewallet.py is not adapted for Dash Core yet.")
```

File: test/get_previous_releases.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+
+def check_host(args) -> int:
+    args.host = os.environ.get('HOST', subprocess.check_output(
+        './depends/config.guess').decode())
+    if args.download_binary:
+        platforms = {
+            'aarch64-*-linux*': 'aarch64-linux-gnu',
+            'x86_64-*-linux*': 'x86_64-linux-gnu',
+            'x86_64-apple-darwin*': 'osx64',
```
Comment (UdjinM6) [inline test/get_previous_releases.py:215]: pls backport 22790 to make it useable on arm macs too

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
👍 

a couple of suggestions

**Comment by UdjinM6** [comment]
LGTM, will utACK after https://github.com/dashpay/dash/pull/5490 merge

**Comment by knst** [comment]
> This pull request has conflicts, please rebase.

> LGTM, will utACK after https://github.com/dashpay/dash/pull/5490 merge

rebased to top of `develop`

**Review by UdjinM6 — APPROVED** [review]
utACK

File: .travis.yml [hunk 3]
```diff
@@ -31,7 +31,7 @@ services:
 cache:
   ccache: true
   directories:
-    - $BASE_BUILD_DIR/ci/scratch/.ccache
```
Comment (PastaPastaPasta) [inline .travis.yml:34]: is this dropping cache?

**Review by PastaPastaPasta — DISMISSED** [review]
utACK for merging via merge commit; one question to confirm

File: .travis.yml (hunk 3 above)
Comment (UdjinM6) [reply .travis.yml:34]: we don't use Travis

Comment (PastaPastaPasta) [reply .travis.yml:34]: I guess that is true :D

Comment (knst) [reply .travis.yml:34]: We don't use Travis, I'm just keeping this changes for backport compatibility. It doesn't affect any caches bcs no travis at all. But if there's any good idea, we can adapt it for our CI.

File: test/functional/test_framework/test_framework.py [hunk 4]
```diff
@@ -255,6 +256,8 @@ def setup(self):
 
         self.options.previous_releases_path = os.getenv("PREVIOUS_RELEASES_DIR") or os.getcwd() + "/releases"
 
+        self.options.previous_releases_path = os.getenv("PREVIOUS_RELEASES_DIR") or os.getcwd() + "/releases"
+
```
Comment (UdjinM6) [inline test/functional/test_framework/test_framework.py:260]: we already have this it seems, see above
```suggestion
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: .travis.yml (hunk 3 above)
Comment (UdjinM6) [reply .travis.yml:34]: imo it's fine to just backport changes for Travis as is and implement similar dashified ones for Gitlab

what is missing in our backport is release cache they added when they changed this line, we could add one too - pls see (and test) 7fe8c49722

this PR:
first run https://gitlab.com/dashpay/dash/-/jobs/4911762972#L106 (downloading)
second run https://gitlab.com/dashpay/dash/-/jobs/4934280594#L106 (downloading)

7fe8c49722:
first run https://gitlab.com/UdjinM6/dash/-/jobs/4935068647#L59 (downloading)
another branch with a dummy commit on top https://gitlab.com/UdjinM6/dash/-/jobs/4935596048#L62 (using cache)

**Comment by UdjinM6** [comment]
maybe also ad2ab002a9 cause why having unused files/hashes listed there? (I know they have it in bitcoin but it makes no sense to me and we are dashifying the list anyway)

**Comment by knst** [comment]
> @[knst](https://github.com/knst) knst [force-pushed](https://github.com/dashpay/dash/compare/03ed63d2a8fac9efa9466a95c6659bd6a91e4699..f78bde4e7159dd79ccb0be0dbabffb4f829fd934) the bc-bp-v20-missing-8v2 branch from[ 03ed63d ](https://github.com/dashpay/dash/commit/03ed63d2a8fac9efa9466a95c6659bd6a91e4699)to f78bde4

applied suggestions and removed one commits that should not belong to this PR https://github.com/dashpay/dash/commit/62019974f47c34174a8f506cde1ac126707f465e

> maybe also https://github.com/dashpay/dash/commit/ad2ab002a970bf225da85c0c21bd3af63609b3c8 cause why having unused files/hashes listed there? (I know they have it in bitcoin but it makes no sense to me and we are dashifying the list anyway)

I made this list not based on list of files, that bitcoin used, but based on list of hashes that we provide with each release, for example for version [19.3.0](https://github.com/dashpay/dash/releases/tag/v19.3.0) I used this list of hashes:  https://github.com/dashpay/dash/releases/download/v19.3.0/SHA256SUMS.asc
```
a4b555b47f5f9a5a01fc5d3b543731088bd10a65dd7fa81fb552818146e424b5  dashcore-19.3.0-aarch64-linux-gnu.tar.gz
531bb188c1aea808ef6f3533d71182a51958136f6e43d9fcadaef1a5fcdd0468  dashcore-19.3.0-osx.dmg
1b4673a2bd71f9f2b593c2d71386e60f4744b59b57142707f0045ed49c92024b  dashcore-19.3.0-osx64.tar.gz
d23cd59ab3a230ebb9bd34fa6329e0d157ecfdbd133f171dfdfa08039d0b3983  dashcore-19.3.0-riscv64-linux-gnu.tar.gz
b5c1860440f97dbb79b1d79bcc48fb2dcc7f0915dd0c4f9fc77aba9cab0294f3  dashcore-19.3.0-win64-setup.exe
8a288189bd4b7c23bb1f917256290dd606d9d47a533dcede0c6190a8f4722e1a  dashcore-19.3.0-win64.zip
c2f3ff5631094abe16af8e476d1197be8685ee20601deda5cad0c34fc879c3de  dashcore-19.3.0-x86_64-linux-gnu.tar.gz
b4bb6bec21213e47586607657e69b0a53905e3c32e2e8e650e93db54dce572d8  dashcore-19.3.0.tar.gz
```

Instead choosing which exactly files is used for which version I just included all of them for any possible further changes and whatever new file would be required by further changes in related scripts. So, I don't feel like removing some hashes from list is "dashyfication", it's more for me as an extra step of clean-up used data.
@UdjinM6 , please, tell me if you still want me to remove all hashes that are not used.

**Comment by knst** [comment]
@PastaPastaPasta when merging this PR, please, include this text to merge-commit message (all a PR description):

-------------
All these backports are bringing next changes:

-  new functional test mempool_compatibility.py
-  new functional test feature_backwards_compatibility.py
-  enables backported before functional test wallet_upgradewallet.py
-  utility script test/get_previous_releases.py to get binaries with previous version

Although, wallet_upgradewallet.py is disabled intentionally - it's not adopted for dash specific so far as HD wallets are not enabled by default atm.

Steps to run new functional tests locally:
```
$ test/get_previous_releases.py -b v19.3.0 v18.2.2 v0.17.0.3 v0.16.1.1 v0.15.0.0
$ test/functional/test_runner.py feature_backwards_compatibility.py mempool_compatibility.py wallet_upgradewallet.py
feature_backwards_compatibility.py | ✓ Passed  | 15 s
mempool_compatibility.py           | ✓ Passed  | 4 s
wallet_upgradewallet.py            | ○ Skipped | 1 s
ALL                                | ✓ Passed  | 20 s (accumulated) 
```

**Review by UdjinM6 — DISMISSED** [review]
utACK

(NOTE: I restarted `linux64_tsan-test` a couple of times already and it keeps failing... these failure aren't related to this PR specifically I think but we might want to take a closer look at them)

**Comment by knst** [comment]
> @[knst](https://github.com/knst) knst [force-pushed](https://github.com/dashpay/dash/compare/f78bde4e7159dd79ccb0be0dbabffb4f829fd934..59d64a9f7f51c65dcbe56e34315033f31f279b7c) the bc-bp-v20-missing-8v2 branch from[ f78bde4 ](https://github.com/dashpay/dash/commit/f78bde4e7159dd79ccb0be0dbabffb4f829fd934)to 59d64a9

and 

@[knst](https://github.com/knst) knst [force-pushed](https://github.com/dashpay/dash/compare/59d64a9f7f51c65dcbe56e34315033f31f279b7c..582f1d466362eb769a3b455321b6b0ebff5d3e39) the bc-bp-v20-missing-8v2 branch from[ 59d64a9 ](https://github.com/dashpay/dash/commit/59d64a9f7f51c65dcbe56e34315033f31f279b7c)to 582f1d4

to rebased on top of newer commit in develop

**Review by UdjinM6 — APPROVED** [review]
LGTM, utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Pull request merged**